### PR TITLE
Add support for precompiled subroutines

### LINKED ISSUE #19: Add support for precompiled subroutines
The SDK should provide an interface for 
- submitting custom NetQASM subroutines to the quantum node controller.
- changing the custom subroutine contents based on runtime information

### PR DESCRIPTION
**Pull request by bvdvecht**
Closes #19

### CODE REVIEW

File: netqasm/__init__.py [hunk 1]
```diff
@@ -2,4 +2,4 @@
 __version__ = "0.9.0"
 
 # Version of the language
-NETQASM_VERSION = (0, 0)
+NETQASM_VERSION = (0, 9)
```
Comment (QFer) [inline netqasm/__init__.py]: Actually you already released 0.9.0. So this is 0.9.1 or higher (it cannot be 0.9.0)

File: netqasm/lang/instr/base.py [hunk 2]
```diff
@@ -207,13 +207,17 @@ def serialize(self) -> bytes:
...
+    def from_operands(cls, operands: List[Union[Operand, int]]):
         assert len(operands) == 3
         reg, imm0, imm1 = operands
         assert isinstance(reg, Register)
-        assert isinstance(imm0, int)
-        assert isinstance(imm1, int)
-        return cls(reg=reg, imm0=Immediate(value=imm0), imm1=Immediate(value=imm1))
+        assert isinstance(imm0, int) or isinstance(imm0, Immediate)
+        if isinstance(imm0, int):
+            imm0 = Immediate(value=imm0)
```
Comment (QFer) [inline netqasm/lang/instr/base.py]: Now imm0 (and imm1) is assigned twice, is mypy happy with that? In the end imm0 must be instance of Immediate. You may rewrite to 1 assignment.

imm0 = Immediate(operands[1]) if isinstance(operands[1], int) else operands[1]
assert isinstance(imm0, Immediate)

File: netqasm/lang/ir.py [hunk 3]
```diff
@@ -120,14 +121,7 @@ def string_to_instruction(instr_str):
...
 
-T_OperandUnion = Union[
-    int,
-    Register,
-    Address,
-    ArrayEntry,
-    ArraySlice,
-    Label,
-]
+T_ProtoOperand = Union[int, Label, Operand]
```
Comment (QFer) [inline netqasm/lang/ir.py]: Maybe define these types in the top of the file when used in multiple other python files. To find them easily.

File: netqasm/lang/ir.py [hunk 4]
```diff
@@ -196,25 +190,103 @@ def _build_str(self, show_lineno=False):
...
+        arguments: Optional[List[str]] = None,
+        netqasm_version: Tuple[int, int] = NETQASM_VERSION,
+        app_id: Optional[int] = None,
+    ) -> None:
+        self._netqasm_version: Tuple[int, int] = netqasm_version
+        self._app_id: Optional[int] = app_id
+
+        self._commands: List[Union[ICmd, BranchLabel]] = []
+        if commands is not None:
+            self._commands = commands
```
Comment (QFer) [inline netqasm/lang/ir.py]: This one doesn't use the setter you defined

File: netqasm/lang/ir.py [hunk 5]
```diff
@@ -196,25 +190,103 @@ def _build_str(self, show_lineno=False):
...
+    def arguments(self) -> List[str]:
+        return self._arguments
 
     def __str__(self):
-        to_return = f"PreSubroutine (netqasm_version={self.netqasm_version}, app_id={self.app_id}):\n"
-        to_return += " LN | HLN | CMD\n"
+        result = "ProtoSubroutine"
+
+        if len(self.arguments) > 0:
+            result += "("
```
Comment (QFer) [inline netqasm/lang/ir.py]: more pythonic is:
result = '(' + ','.join(arg_name for arg_name in self.arguments) + ')'

File: netqasm/lang/ir.py [hunk 6]
```diff
@@ -196,25 +190,103 @@ def _build_str(self, show_lineno=False):
...
+            ops: List[T_ProtoOperand] = []
+            for op in cmd.operands:
+                if isinstance(op, Template):
+                    ops.append(arguments[op.name])
+                else:
+                    ops.append(op)
+            cmd.operands = ops
+            commands.append(cmd)
+
+        self._commands = commands
```
Comment (QFer) [inline netqasm/lang/ir.py]: also this one should use the setter to concentrate the setting of _commands to 1 spot

File: netqasm/lang/subroutine.py [hunk 7]
```diff
@@ -8,47 +8,127 @@
...
+        arguments: Optional[List[str]] = None,
+        netqasm_version: Tuple[int, int] = NETQASM_VERSION,
+        app_id: Optional[int] = None,
+    ) -> None:
+        self._netqasm_version: Tuple[int, int] = netqasm_version
+        self._app_id: Optional[int] = app_id
+
+        self._instructions: List[NetQASMInstruction] = []
+        if instructions is not None:
+            self._instructions = instructions
```
Comment (QFer) [inline netqasm/lang/subroutine.py]: The same as for commands earlier, use the setter

File: netqasm/lang/subroutine.py [hunk 8]
```diff
@@ -8,47 +8,127 @@
...
+            ops: List[Union[Operand, int]] = []
+            for op in instr.operands:
+                if isinstance(op, Template):
+                    assert arguments is not None
+                    ops.append(arguments[op.name])
+                else:
+                    ops.append(op)
+            instrs.append(instr.from_operands(ops))
+
+        self._instructions = instrs
```
Comment (QFer) [inline netqasm/lang/subroutine.py]: idem, better use the setter

File: netqasm/lang/subroutine.py [hunk 9]
```diff
@@ -8,47 +8,127 @@
...
 
     def __str__(self):
-        to_return = f"Subroutine (netqasm_version={self.netqasm_version}, app_id={self.app_id}):\n"
-        to_return += " LN | HLN | CMD\n"
-        for i, command in enumerate(self.commands):
-            if isinstance(command, DebugInstruction):
-                to_return += f"# {command.text}\n"
+        result = "Subroutine"
+        if len(self.arguments) > 0:
+            result += "("
```
Comment (QFer) [inline netqasm/lang/subroutine.py]: idem as earlier about string concatenation with join

File: tests/test_external/test_sdk/test_precompilation.py [hunk 10]
```diff
@@ -0,0 +1,62 @@
+from netqasm.lang.operand import Template
```
Comment (QFer) [inline tests/test_external/test_sdk/test_precompilation.py:1]: This file has great resemblance with examples/sdk_scripts/precompile.py
I guess the purpose is different.

File: tests/test_subroutine.py [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+        assert arg1 == arg2
+
+
+def test_parse_and_instantiate():
+    template = """
+rot_z R0 {num} {denom}
+"""
+
+    expected_text = """
+# NETQASM 0.0
```
Comment (QFer) [inline tests/test_subroutine.py]: I saw earlier a  version change to 0.9

**Review by QFer — CHANGES_REQUESTED** [review]

File: netqasm/__init__.py (hunk 1 above)
Comment (bvdvecht) [reply netqasm/__init__.py]: Updated to (0, 10). (The NETQASM_VERSION does not have to be in sync with the package version. Also, until now this NETQASM_VERSION has never really been updated, and also not really used). I thought that while updating it now for the first time, I might as well use the current package version. But it's not quite clear yet to me what exact protocol to follow in updating this NETQASM_VERSION number.

File: netqasm/lang/instr/base.py (hunk 2 above)
Comment (bvdvecht) [reply netqasm/lang/instr/base.py]: Implemented your suggestion

File: netqasm/lang/ir.py (hunk 3 above)
Comment (bvdvecht) [reply netqasm/lang/ir.py]: Done

File: netqasm/lang/ir.py (hunk 4 above)
Comment (bvdvecht) [reply netqasm/lang/ir.py]: Updated

File: netqasm/lang/ir.py (hunk 5 above)
Comment (bvdvecht) [reply netqasm/lang/ir.py]: Implemented your suggestion

File: netqasm/lang/ir.py (hunk 6 above)
Comment (bvdvecht) [reply netqasm/lang/ir.py]: Done

File: netqasm/lang/subroutine.py (hunk 7 above)
Comment (bvdvecht) [reply netqasm/lang/subroutine.py]: Done

File: netqasm/lang/subroutine.py (hunk 8 above)
Comment (bvdvecht) [reply netqasm/lang/subroutine.py]: Done

File: netqasm/lang/subroutine.py (hunk 9 above)
Comment (bvdvecht) [reply netqasm/lang/subroutine.py]: Done

File: tests/test_external/test_sdk/test_precompilation.py (hunk 10 above)
Comment (bvdvecht) [reply tests/test_external/test_sdk/test_precompilation.py:1]: You're right. Removed `precompile.py`.

File: tests/test_subroutine.py (hunk 11 above)
Comment (bvdvecht) [reply tests/test_subroutine.py]: Updated to use NETQASM_VERSION

**Review by QFer — APPROVED** [review]

**Pull request merged**